File: recomment.php
<?php
    
    /* Remove comments from a program*/

    $handle = @fopen("redis-2.6.17/src/intset.c", 'r');

    if (!$handle)
    {
        echo "Open file error\n";
        exit;
    }

    while (($v = fgetc($handle)) !== false)
        delcomment($v, $handle);

    fclose($handle);

    function delcomment($v1, $handle)
    {
        if ($v1 == '/')
        {
            $v2 = fgetc($handle);

            if ($v2 == '*')
            {
                delmulitLine($handle);
            }
            else if ($v2 == '/')
            {
                fgets($handle);
            }
            else
            {
                echo $v1 . $v2;
            }
        }
        else
        {
            echo $v1;
        }
    }

    function delmulitLine($handle)
    {
        $c1 = fgetc($handle);
        $c2 = fgetc($handle);

        while (($c1 != '*') || ($c2 != '/'))
        {
            $c1 = $c2;
            $c2 = fgetc($handle);
        }
    }

File: recomment.c
/* Remove comments from a program */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define FILE_MAX_CHAR 1024 * 1024
#define LINE_MAX_CHAR 1024

void delmulitLine(char *p);
void delLine(char *p);

char * recomment(char *str)
{
    char *p = str;
    char *pval = (char *)malloc(sizeof(char) * FILE_MAX_CHAR);
    char *pval1 = pval;

    do
    {
        if (*p == '/')
        {
            p++;

            if (*p == '*')
            {
                delmulitLine(p);
            }
            else if (*p == '/')
            {
                delLine(p);
            }
            else
            {
                *(++pval1) = '/';
                *(++pval1) = *p;
            }
        }
        else
        {
            *(++pval1) = *p;
        }

    }while(p++);

    return pval;
}

void delmulitLine(char *p)
{
    char p1 = *++p;
    char p2 = *++p;

    while ((p1 != '*') || (p2 != '/'))
    {
        printf("%c", p1);
        p1 = p2;
        p2 = *(++p);
    }
}

void delLine(char *p)
{
    while (*p++ != '\n');
}

int main()
{
    char *file = "redis-2.6.17/src/intset.c";
    FILE *pFile = fopen(file, "r");
    char content[FILE_MAX_CHAR] = {0};
    char *pcontent = content;

    char line[LINE_MAX_CHAR] = {0};
    char *pline = line;

    if (pFile == NULL)
    {
        printf("Open file error");
        return 1;
    }

    int lastlen = 0;

    while ((fgets(pline, LINE_MAX_CHAR, pFile)) != NULL)
    {
        int len = strlen(pline);
        memcpy(pcontent+lastlen, pline, len);
        memset(pline, 0, LINE_MAX_CHAR);
        lastlen += len;
    }

    fclose(pFile);
    
    char *pval = recomment(content);
    printf("%s\n", pval);

    return 0;
}
